sql: expose mvcc timestamps to SQL

rohany opened the pull request:
Fixes #50102.

This PR adds introspection into the KV layer's concurrency control from
the SQL level. In particular, we now expose the MVCC HLC timestamp of a
row as a special system column on every table. This system column is
exposed as a decimal, and is computed as `wall time * 10^10 + logical time`.

To accomplish this, this PR adds planning and execution infrastructure
for implicit system columns that can be produced by the execution layer
for a particular row.

Release note (sql change): Expose the MVCC timestamp of each row as a
system column on tables. This column is named
`crdb_internal_mvcc_timestamp` and is accessible only in a limited set
of contexts.

rohany committed:
sql: expose mvcc timestamps to SQL

Fixes #50102.

This PR adds introspection into the KV layer's concurrency control from
the SQL level. In particular, we now expose the MVCC HLC timestamp of a
row as a special system column on every table. This system column is
exposed as a decimal, and is computed as `wall time * 10^10 + logical time`.

To accomplish this, this PR adds planning and execution infrastructure
for implicit system columns that can be produced by the execution layer
for a particular row.

Release note (sql change): Expose the MVCC timestamp of each row as a
system column on tables. This column is named
`crdb_internal_mvcc_timestamp` and is accessible only in a limited set
of contexts.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/51494)
<!-- Reviewable:end -->

rohany commented on pkg/sql/logictest/testdata/logic_test/mvcc:
```diff
@@ -0,0 +1,124 @@
+statement error pq: INSERT has more expressions than target columns, 4 expressions for 3 targets
+INSERT INTO t VALUES (7, 8, 9, 1.0)
+
+# TODO (rohany): I'm not sure how to disable returning the system columns. The way
```
I'll need some pointers here.

rohany commented on pkg/sql/opt/exec/execbuilder/testdata/select_for_update:
```diff
@@ -525,78 +525,40 @@ root                 ·                 ·
- └── scan   ·                   ·
-·           table               t@primary
-·           spans               FULL SCAN
+·     distribution      local
```
I'm not sure why these changed, my changes shouldn't have affected the plans like this. Though the plan does seem reasonable since the in is a tautology?

rohany commented on pkg/sql/opt/norm/prune_cols_funcs.go:
```diff
@@ -184,6 +184,16 @@ func neededMutationFetchCols(
+	// primary index. These columns are implicit and should not be part of the set
+	// of fetch columns, so we remove them here.
+	table := tabMeta.Table
+	for i := 0; i < table.ColumnCount(); i++ {
```
Because of the API change, there isn't an easy way to iterate over only the system columns of a table. I'm not sure that is a big deal though.

rohany commented:
cc @RaduBerinde @andy-kimball 

I've updated this to be on top of radu's changes. I've omitted the commit on top of this that adds the system column to the test catalog and updates the various opt tests. I went through all uses of `ColumnCount` and made a decision about whether to include the system column or not. I'll need some double checking about whether those decisions were correct + how to write tests that tickle those specific pieces of code.

RaduBerinde reviewed (commented):
Thanks! Some initial comments about `cat`

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51494#-:-MCO9bM1DBX2s7s_nh4r:b64f2zq)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rohany)

___
*[pkg/sql/opt/cat/column.go, line 32 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/51494#-MCO94rh4hSfMoBg7M-h:-MCO94rh4hSfMoBg7M-i:bb5kbfp) ([raw file](https://github.com/cockroachdb/cockroach/blob/d77df5de4c956208efe1c631ce77f43c9c0ce564/pkg/sql/opt/cat/column.go#L32)):*
> ```Go
> 	DeleteOnly
> 	// SystemColumn columns are implicit columns that every physical table
> 	// contains. These columns can only be read from and must not be included
> ```

Add that they cannot be part of the key cols / lax key cols for indexes.
___
*[pkg/sql/opt/cat/column.go, line 33 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/51494#-MCO1lJ94qjqVAsbT9R6:-MCO1lJ94qjqVAsbT9R7:b-plmngf) ([raw file](https://github.com/cockroachdb/cockroach/blob/d77df5de4c956208efe1c631ce77f43c9c0ce564/pkg/sql/opt/cat/column.go#L33)):*
> ```Go
> 	// SystemColumn columns are implicit columns that every physical table
> 	// contains. These columns can only be read from and must not be included
> 	// as part of mutations. These columns are synthesized entirely within the
> ```

"synthesized entirely within the optimizer" is confusing. They are synthesized in the catalog implementation, not the optimizer (and their values are also synthesized outside the optimizer). I would just keep this detail out, it's not relevant to comprehending the interface.
___
*[pkg/sql/opt/cat/column.go, line 35 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/51494#-MCO1d_Y8vLVs8Bu6-hh:-MCO1d_Z9328cBAj88YN:b-drk159) ([raw file](https://github.com/cockroachdb/cockroach/blob/d77df5de4c956208efe1c631ce77f43c9c0ce564/pkg/sql/opt/cat/column.go#L35)):*
> ```Go
> 	// as part of mutations. These columns are synthesized entirely within the
> 	// optimizer and are not part of the persisted table definition.
> 	SystemColumn
> ```

[nit] the other enums don't have `Column`. I'd keep just as `System` even if it's kind of generic
___
*[pkg/sql/opt/cat/index.go, line 58 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/51494#-MCO2Tbz8LRGCeq0YWnp:-MCO2Tbz8LRGCeq0YWnq:bowvew) ([raw file](https://github.com/cockroachdb/cockroach/blob/d77df5de4c956208efe1c631ce77f43c9c0ce564/pkg/sql/opt/cat/index.go#L58)):*
> ```Go
> 	// excludes any system columns. These columns shouldn't be considered by
> 	// statistics when computing the cost of scanning an index.
> 	ColumnCountNoSystemColumns() int
> ```

I'd remove this if possible. Looks like we are using it in just one place (in the coster). We could easily go through the columns and check how many are system columns. I think that code will one day have to look at each column's average value size (or something like that) anyway.


<!-- Sent from Reviewable.io -->

rohany reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51494)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RaduBerinde)

___
*[pkg/sql/opt/cat/column.go, line 32 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/51494#-MCO94rh4hSfMoBg7M-h:-MCOD3s8DbHVVRV49Byp:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/d77df5de4c956208efe1c631ce77f43c9c0ce564/pkg/sql/opt/cat/column.go#L32)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Add that they cannot be part of the key cols / lax key cols for indexes.
</blockquote></details>

Done
___
*[pkg/sql/opt/cat/column.go, line 33 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/51494#-MCO1lJ94qjqVAsbT9R6:-MCOD4Ko7Nu_bI9N59X7:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/d77df5de4c956208efe1c631ce77f43c9c0ce564/pkg/sql/opt/cat/column.go#L33)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

"synthesized entirely within the optimizer" is confusing. They are synthesized in the catalog implementation, not the optimizer (and their values are also synthesized outside the optimizer). I would just keep this detail out, it's not relevant to comprehending the interface.
</blockquote></details>

Done
___
*[pkg/sql/opt/cat/column.go, line 35 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/51494#-MCO1d_Y8vLVs8Bu6-hh:-MCOD5vO8BfcurXGHB6o:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/d77df5de4c956208efe1c631ce77f43c9c0ce564/pkg/sql/opt/cat/column.go#L35)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] the other enums don't have `Column`. I'd keep just as `System` even if it's kind of generic
</blockquote></details>

Done
___
*[pkg/sql/opt/cat/index.go, line 58 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/51494#-MCO2Tbz8LRGCeq0YWnp:-MCOD6bf43S1H1OkRSnw:b-xhmemi) ([raw file](https://github.com/cockroachdb/cockroach/blob/d77df5de4c956208efe1c631ce77f43c9c0ce564/pkg/sql/opt/cat/index.go#L58)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I'd remove this if possible. Looks like we are using it in just one place (in the coster). We could easily go through the columns and check how many are system columns. I think that code will one day have to look at each column's average value size (or something like that) anyway.
</blockquote></details>

I added this because I didn't know if it was acceptable in the coster to iterate over all of the columns in the index. If that's fine i wont add this.


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
In cat.Table we should document how system columns interact with column families (I assume they don't show up in any family?)

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51494#-:-MCP-wP621_mSzHWOfbF:binxc6)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rohany)

___
*[pkg/sql/logictest/testdata/logic_test/mvcc, line 111 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51494#-MCKn91qzaZV8_DSmd7y-r1-111:-MCP2Omv3fHsYm0pZrG2:b-5zw43e) ([raw file](https://github.com/cockroachdb/cockroach/blob/842cb0f25880dd4869aace189c8d65bb886f364a/pkg/sql/logictest/testdata/logic_test/mvcc#L111)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

I'll need some pointers here.
</blockquote></details>

I think in `mutationBuilder.buildReturning`, we'd want that `inScope.appendColumnsFromTable` call inside to skip system columns (just like we skip mutation columns)
___
*[pkg/sql/opt/cat/index.go, line 57 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/51494#-MCP-COq1p19CdKEIQH0:-MCP-COrAE72A5BEJiWg:bwo2jj3) ([raw file](https://github.com/cockroachdb/cockroach/blob/8fd30acd6b8bd21578d7b75ddd0085790bf73434/pkg/sql/opt/cat/index.go#L57)):*
> ```Go
> 	// ColumnCount returns the number of columns in the index. This includes
> 	// columns that were part of the index definition (including the STORING
> 	// clause), as well as implicitly added primary key columns. This also the
> ```

remove "also the"
___
*[pkg/sql/opt/exec/execbuilder/testdata/select_for_update, line 528 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51494#-MCN1rsQnv1H5F3eWnsw-r1-528:-MCP34YL-XsYZtvkthb3:b-o4bfk7) ([raw file](https://github.com/cockroachdb/cockroach/blob/842cb0f25880dd4869aace189c8d65bb886f364a/pkg/sql/opt/exec/execbuilder/testdata/select_for_update#L528)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

I'm not sure why these changed, my changes shouldn't have affected the plans like this. Though the plan does seem reasonable since the in is a tautology?
</blockquote></details>

Not sure either.. I'll try to figure it out. In any case, maybe change to `IN (SELECT b FROM t)`. Then it shouldn't be correct to drop the join
___
*[pkg/sql/opt/norm/prune_cols_funcs.go, line 191 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51494#-MCKn91o79z0jLvT7BjY-r1-191:-MCP0IY8EbX7mXW3aKHD:b-m51tll) ([raw file](https://github.com/cockroachdb/cockroach/blob/842cb0f25880dd4869aace189c8d65bb886f364a/pkg/sql/opt/norm/prune_cols_funcs.go#L191)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

Because of the API change, there isn't an easy way to iterate over only the system columns of a table. I'm not sure that is a big deal though.
</blockquote></details>

I think this would be more clear if we just did this check above instead of `cols.UnionWith(indexCols)`. I think that's the only way a system column would make its way into this set in the first place.


<!-- Sent from Reviewable.io -->

rohany reviewed (commented):
Done, added a blurb on the `System` columnkind

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51494#-:-MCPr2mg785MY1RTH1uY:ba10r6w)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RaduBerinde)

___
*[pkg/sql/logictest/testdata/logic_test/mvcc, line 111 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51494#-MCKn91qzaZV8_DSmd7y-r1-111:-MCPr56V9bkmCKQacxWa:b-j5wfbx) ([raw file](https://github.com/cockroachdb/cockroach/blob/842cb0f25880dd4869aace189c8d65bb886f364a/pkg/sql/logictest/testdata/logic_test/mvcc#L111)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I think in `mutationBuilder.buildReturning`, we'd want that `inScope.appendColumnsFromTable` call inside to skip system columns (just like we skip mutation columns)
</blockquote></details>

That worked. I think that it is reasonable to skip system columns in `appendColumnsFromTable` so we don't accidentally add system columns.
___
*[pkg/sql/opt/cat/index.go, line 57 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/51494#-MCP-COq1p19CdKEIQH0:-MCPrCUCBze2AdSmjfya:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/8fd30acd6b8bd21578d7b75ddd0085790bf73434/pkg/sql/opt/cat/index.go#L57)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

remove "also the"
</blockquote></details>

done
___
*[pkg/sql/opt/exec/execbuilder/testdata/select_for_update, line 528 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51494#-MCN1rsQnv1H5F3eWnsw-r1-528:-MCPrDK-6Ad5tkd-Uotq:baxmvar) ([raw file](https://github.com/cockroachdb/cockroach/blob/842cb0f25880dd4869aace189c8d65bb886f364a/pkg/sql/opt/exec/execbuilder/testdata/select_for_update#L528)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Not sure either.. I'll try to figure it out. In any case, maybe change to `IN (SELECT b FROM t)`. Then it shouldn't be correct to drop the join
</blockquote></details>

We produce joins with the select b from t.
___
*[pkg/sql/opt/norm/prune_cols_funcs.go, line 191 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51494#-MCKn91o79z0jLvT7BjY-r1-191:-MCPrICB6axg4YD4yBCD:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/842cb0f25880dd4869aace189c8d65bb886f364a/pkg/sql/opt/norm/prune_cols_funcs.go#L191)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I think this would be more clear if we just did this check above instead of `cols.UnionWith(indexCols)`. I think that's the only way a system column would make its way into this set in the first place.
</blockquote></details>

Done


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51494)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RaduBerinde and @rohany)

___
*[pkg/sql/logictest/testdata/logic_test/mvcc, line 111 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51494#-MCKn91qzaZV8_DSmd7y-r1-111:-MCSYtda2WcTyBAW6FcX:b-jdnpkd) ([raw file](https://github.com/cockroachdb/cockroach/blob/842cb0f25880dd4869aace189c8d65bb886f364a/pkg/sql/logictest/testdata/logic_test/mvcc#L111)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

That worked. I think that it is reasonable to skip system columns in `appendColumnsFromTable` so we don't accidentally add system columns.
</blockquote></details>

Looks like that would be correct for all call sites, maybe rename to `appendOrdinaryColumnsFromTable`


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
___
*[pkg/sql/opt/exec/execbuilder/testdata/select_for_update, line 528 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51494#-MCN1rsQnv1H5F3eWnsw-r1-528:-MCSnD3u4FbqTReJtG8P:bisl0lk) ([raw file](https://github.com/cockroachdb/cockroach/blob/842cb0f25880dd4869aace189c8d65bb886f364a/pkg/sql/opt/exec/execbuilder/testdata/select_for_update#L528)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

We produce joins with the select b from t.
</blockquote></details>

I've been trying to figure this out. I can't find a rule that would do this transformation (though we'll probably add it in the future), so something fishy must be going on. I can't use `optsteps` reproduce inside opt tests without the testcatalog change. I've been trying to hack the planning code but it's hard. I can try again once you get the testcatalog working too.


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
___
*[pkg/sql/opt/exec/execbuilder/testdata/select_for_update, line 528 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51494#-MCN1rsQnv1H5F3eWnsw-r1-528:-MCSxHWE49rsOxZtvMYk:b5y3gxb) ([raw file](https://github.com/cockroachdb/cockroach/blob/842cb0f25880dd4869aace189c8d65bb886f364a/pkg/sql/opt/exec/execbuilder/testdata/select_for_update#L528)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I've been trying to figure this out. I can't find a rule that would do this transformation (though we'll probably add it in the future), so something fishy must be going on. I can't use `optsteps` reproduce inside opt tests without the testcatalog change. I've been trying to hack the planning code but it's hard. I can try again once you get the testcatalog working too.
</blockquote></details>

Ah, I think I figured it out. We have a rule `EliminateJoinUnderProjectLeft` that would do this if there is a Project on top. With your change, we have a Project to eliminate the system column. I was able to get the same plan on `master` with `SELECT a` instead of `SELECT *`. So it's all good.


<!-- Sent from Reviewable.io -->

rohany commented:
> Ah, I think I figured it out. We have a rule EliminateJoinUnderProjectLeft that would do this if there is a Project on top. With your change, we have a Project to eliminate the system column. I was able to get the same plan on master with SELECT a instead of SELECT *. So it's all good.

Nice. Why does that rule need a projection to fire though?

RaduBerinde commented:
> Nice. Why does that rule need a projection to fire though?

We are just missing a corresponding (simpler) rule when there is no projection. The rule with the projection does more things than eliminate an unnecessary semi-join.

rohany commented:
that makes sense. @RaduBerinde do you think it's reasonable for me to start updating the opt tests, or there are more things here that you want to change?

RaduBerinde commented:
Let me look over it more, I didn't get a chance to look at all the changes.

RaduBerinde reviewed (commented):
opt part LGTM. Unfortunately you will have to merge with Andy's changes in 51574.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51494#-:-MCgewSXErCk9FSCeFlU:b-g8mvz9)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rohany)

___
*[pkg/sql/opt/cat/utils.go, line 287 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/51494#-MCgesn75tKm_Z4QegOq:-MCgesn75tKm_Z4QegOr:b-ryh5gx) ([raw file](https://github.com/cockroachdb/cockroach/blob/cecb2fe0f22e22e78a3fd8bde60c66e3ddb98de8/pkg/sql/opt/cat/utils.go#L287)):*
> ```Go
> }
> 
> func formatColumn(col Column, isMutationCol bool, isSystemCol bool, buf *bytes.Buffer) {
> ```

[nit] should just pass the `cat.ColumnKind`


<!-- Sent from Reviewable.io -->

rohany committed:
opt: update opt test catalog to include system column and update tests

This commit adds the MVCC timestamp system column to the test catalog
and updates all of the tests with the change.

Release note: None

rohany commented:
I rebased on top of andy's change by adding a `system` bool to the `scopeColumn` and not adding those columns to the `fetchColIDs` list. Let me know if that doesn't sound ideal.

I'll begin editing the opt tests as CI comes in.

RaduBerinde reviewed (commented):
LGTM, just some minor nits

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51494#-:-MCh4EiN7h9A1YhE5v3M:bljgxka)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rohany)

___
*[pkg/sql/opt/optbuilder/fk_cascade.go, line 116 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/51494#-MCh2sjv9i68ZK3EECto:-MCh2sjv9i68ZK3EECtp:bpile84) ([raw file](https://github.com/cockroachdb/cockroach/blob/55193b5c7ed7f9d0745cdc791aefa0fe9200b4b3/pkg/sql/opt/optbuilder/fk_cascade.go#L116)):*
> ```Go
> 	// Set list of columns that will be fetched by the input expression.
> 	for i := range mb.outScope.cols {
> 		// Ensure that we don't add system columns to the fetch ords.
> ```

[nit] "fetch ords" is no longer a thing, maybe "fetch columns" (+2 more times in this file)
___
*[pkg/sql/opt/optbuilder/insert.go, line 521 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/51494#-MCh3E2X4nVqe2LMbzAq:-MCh3E2X4nVqe2LMbzAr:b-31ly8v) ([raw file](https://github.com/cockroachdb/cockroach/blob/55193b5c7ed7f9d0745cdc791aefa0fe9200b4b3/pkg/sql/opt/optbuilder/insert.go#L521)):*
> ```Go
> 	for i, n := 0, mb.tab.ColumnCount(); i < n && numCols < maxCols; i++ {
> 		// Skip mutation, hidden or system columns.
> 		if cat.IsMutationColumn(mb.tab, i) || mb.tab.Column(i).IsHidden() || cat.IsSystemColumn(mb.tab, i) {
> ```

maybe it's cleaner to check that the kind is `Ordinary` (same below)
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 244 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/51494#-MCh3hpQBoDEuq3V-DUZ:-MCh3hpQBoDEuq3V-DU_:b6rxuh4) ([raw file](https://github.com/cockroachdb/cockroach/blob/55193b5c7ed7f9d0745cdc791aefa0fe9200b4b3/pkg/sql/opt/optbuilder/mutation_builder.go#L244)):*
> ```Go
> 	// Set list of columns that will be fetched by the input expression.
> 	for i := range mb.outScope.cols {
> 		// Ensure that we don't add system columns to the fetch ords.
> ```

"fetch columns"
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 371 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/51494#-MCh3jesDxwdRHwfF2ff:-MCh3jesDxwdRHwfF2fg:b6rxuh4) ([raw file](https://github.com/cockroachdb/cockroach/blob/55193b5c7ed7f9d0745cdc791aefa0fe9200b4b3/pkg/sql/opt/optbuilder/mutation_builder.go#L371)):*
> ```Go
> 	// Set list of columns that will be fetched by the input expression.
> 	for i := range mb.outScope.cols {
> 		// Ensure that we don't add system columns to the fetch ords.
> ```

"fetch columns"
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 836 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/51494#-MCh3nkJ1VW2xDr2hWkh:-MCh3nkJ1VW2xDr2hWki:bk3o264) ([raw file](https://github.com/cockroachdb/cockroach/blob/55193b5c7ed7f9d0745cdc791aefa0fe9200b4b3/pkg/sql/opt/optbuilder/mutation_builder.go#L836)):*
> ```Go
> 		private.ReturnCols = make(opt.ColList, mb.tab.ColumnCount())
> 		for i, n := 0, mb.tab.ColumnCount(); i < n; i++ {
> 			if cat.IsMutationColumn(mb.tab, i) || cat.IsSystemColumn(mb.tab, i) {
> ```

[nit] kind == Ordinary
___
*[pkg/sql/opt/optbuilder/scope.go, line 230 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/51494#-MCh3uyv9gRDIFIrsq-G:-MCh3uywAb1xqTn9jpUC:b-jodzzg) ([raw file](https://github.com/cockroachdb/cockroach/blob/55193b5c7ed7f9d0745cdc791aefa0fe9200b4b3/pkg/sql/opt/optbuilder/scope.go#L230)):*
> ```Go
> 	}
> 	for i, n := 0, tab.ColumnCount(); i < n; i++ {
> 		if cat.IsMutationColumn(tab, i) || cat.IsSystemColumn(tab, i) {
> ```

Kind == Ordinary (I'd definitely do this here where the function name specifically mentions Ordinary)
___
*[pkg/sql/opt/optbuilder/scope_column.go, line 48 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/51494#-MCh3Rx-4UADOVHsFZ9w:-MCh3Rx-4UADOVHsFZ9x:bbbk63f) ([raw file](https://github.com/cockroachdb/cockroach/blob/55193b5c7ed7f9d0745cdc791aefa0fe9200b4b3/pkg/sql/opt/optbuilder/scope_column.go#L48)):*
> ```Go
> 	// system is true if the column is an implicit system column. It should not
> 	// be included in mutations.
> 	system bool
> ```

[nit] We could also store the ColumnKind here instead. Not sure if that is better or worse though, your call


<!-- Sent from Reviewable.io -->

rohany commented:
> [nit] "fetch ords" is no longer a thing, maybe "fetch columns" (+2 more times in this file)

Done woops.

Updated all of these checks to check `!= cat.Ordinary`

> [nit] We could also store the ColumnKind here instead. Not sure if that is better or worse though, your call

I think this one is easier to keep as the booleans.

rohany commented:
Alright, I've got the LGTM from @RaduBerinde. Going to merge this now!

bors r=raduberinde
